### DISCUSSION TITLE
Next.js dev and prod build stuck in infinite loop since 2.4.2

### DISCUSSION THREAD

**Discussion by berndartmueller** [ben-rogerson/twin.macro#463] [Q&A]
Hey guys,

I'm having a hard time figuring out why my project (Next.js 10.1.4-canary.10, Preact, linaria) build is stuck in an infinite loop since upgrading from twin.macro 2.4.1 to 2.4.2.

Does someone else have similar issues? 

Thanks a lot!

**Comment by ben-rogerson** [comment]
Are you seeing any console errors that would help track down the bug?

Twin doesn't officially support Linaria yet so I'd need to see a reduced test case to look into this further (no existing examples to test with).

**Reply by berndartmueller** [reply]
Unfortunately, I can't see any errors. But I was able to track the issue down to where it originates.

https://github.com/ben-rogerson/twin.macro/commit/296b81aa573484334fa7dfd4dce0bcd6652d2808#diff-dabc2cc3f8cd8a88bf2cc8ac7cc0b7cd908f0daf7d125d07c89e52372a5bddfdR100

It seems to be caused by the newly introduced `normalizeUserPluginSelectors()` function. If I remove the code which calls this function, it works as usual. 

Even more interestingly, it seems to be caused by the line `return { ...result, ...keys }`. My wild guess is that there is a memory leak.

Here's a CodeSandbox with a stripped down version of how I'm using twin.macro with Linaria: https://codesandbox.io/s/nextjs-preact-ts-twinmacro-linaria-mjo4c

**Reply by berndartmueller** [reply]
Out of curiosity, I rewrote the `normalizeUserPluginSelectors()` function to use a `for` loop instead of `reduce()` and instead of `Object.assign()` I iterate over the keys and directly assign it to an object.

e.g.

```ts
const normalizeUserPluginSelectors = data => {
    const unpackedSelectors = {};

    for (const [selector, value] of Object.entries(data)) {
        const keys = selector
            .split(',')
            .filter(s =>
                isMediaQuery(s) ?
                Object.keys(value).some(selector => isClass(selector)) :
                isClass(s)
            )
            .reduce((result, property) => ({
                ...result,
                [property]: value
            }), {})

        Object.keys(keys).map(key => {
             unpackedSelectors[key] = keys[key];
        })
    }
}
```

Now it works! I suspect that due to the usage of multiple tailwind plugins and a rather large page consisting of many React components, the usage of `reduce()` and `Object.assign()` is not ideal.

**Reply by ben-rogerson** [reply]
Wow, nice find here.
Mind sharing what plugins you're using? So I can test this out myself.

**Reply by berndartmueller** [reply]
I'm using the plugins listed in `tailwind.config.js` in the Codesandbox demo linked above. 

It's just a guess that this issue is caused by having a rather "large" page with many components and many tailwind classes to go through.

But I'm glad if I can help improving twin.macro by reporting this issue :)

**Reply by ben-rogerson** [reply]
hey Bernd, I pulled down that example, fixed the types file and had no issues with the build / dev.
Did you find out what combination of classes would cause the issue?

**Reply by berndartmueller** [reply]
I was not able to replicate the issue with my code example. As I said in my previous comment, it seems to be caused by having a large page with many tailwindcss classes.

Maybe you can replicate it by artificially blowing up the to be processed tailwind classes during the build.

If I can be of any help, let me know!

**Reply by berndartmueller** [reply]
@ben-rogerson  Are you making any progress on this issue? Is there anything that prevents you from rewriting this `normalizeUserPluginSelectors ()` to use a `for` loop instead of `reduce()`?

**Reply by ben-rogerson** [reply]
No progress yet, last time I looked at this I was unable to reproduce the issue.

I loaded up your example with enough plugin classes to bloat the build to 2+ minutes and couldn't reproduce the issue.

Can you perhaps duplicate your project and whittle it down to a minimal reproduction I can test with?

**Reply by berndartmueller** [reply]
Weird.

Ok I’ll try my best to reproduce it in a separate repository.
